2016/exercises/final_project/Alba-Petar-Rebecca-Simon/PolytopeProject/Gale_Diagram.cpp

```diff
@@ -0,0 +1,322 @@
+//
+//  Gale_Diagram.cpp
+//  PolytopeProject
+//
+
+
+#include "Gale_Diagram.hpp"
+
+
+
+Gale_Diagram::Gale_Diagram(vector<Point> points): points(points),matrix(Matrix(points)){
+};
+
+bool Gale_Diagram::is_neighborly() const{
+    //Choose 2 points
+    for(int i=0; i<points.size()-1; i++){
+        for(int j=i+1; j<points.size(); j++){
+            
+            bool edgeIJPresent=false;
+            
+            
+            //Then choose another 2
+            for(int k=0; k<points.size()-1 && (!edgeIJPresent); k++){
+                for(int l=k+1; l<points.size() && (!edgeIJPresent); l++){
+					if( (i!=k) and (i!=l) and (j!=k) and (j!=l)){
+						edgeIJPresent = isSimplex(i, j, k, l);
+					}
+                }
+            }
+            
+            if(!edgeIJPresent){
+                return false;
+            }
+        }
+        
+    }
+    return true;
+}
+
+
+
+vector<int> Gale_Diagram::findMissing(int i, int j, int k, int l) const{
+    vector<int> pointNums(points.size());
+    for(int a=0; a<points.size(); a++) pointNums[a]=a;
+
+	pointNums[i] = -1;
+	pointNums[j] = -1;
+	pointNums[k] = -1;
+	pointNums[l] = -1;
+	
+    vector<int> pN;
+    
+    for(int a=0; a<points.size(); a++){
+		if(pointNums[a]>=0) pN.push_back(a);
+	}
+    
+    return pN;
+}
+
+
+//Does NOT check for degeneracies
+//returns -1 if x lies on the right of the oriented line ab, otherwise +1;
+int Gale_Diagram::orientation(const Point &x, const Point &a, const Point &b) const{
+    Matrix M = Matrix(vector<Point>{b.minus(a),x.minus(a)});
+    return sgn(M.determinant());
+}
+//Checks if onePoint lies in the triangle defined by threePoints
+bool Gale_Diagram::in_triangle(const vector<Point> &onePoint, const vector<Point> &triangle) const{
+    Point x=onePoint[0];
+    Point a=triangle[0];
+    Point b=triangle[1];
+    Point c=triangle[2];
+    int o1 = orientation(x,a,b);
+    int o2 = orientation(x,b,c);
+    int o3 = orientation(x, c,a);
+    return o1==o2 && o2 == o3;
+}
+// returns whether both linesegments (vectors of size 2) intersect.
+bool Gale_Diagram::intersectingLineSegments(const vector<Point> &lineSegment1, const vector<Point> &lineSegment2) const{
+    Point a=lineSegment1[0];
+    Point b=lineSegment1[1];
+    Point c=lineSegment2[0];
+    Point d=lineSegment2[1];
+    int o1 = orientation(a, c, d);
+    int o2 = orientation(b,c,d);
+    int o3 = orientation(c,a,b);
+    int o4 = orientation(d,a,b);
+    return o1!=o2 && o3!=o4;
+}
+bool Gale_Diagram::check_intersecting(const vector<Point> &points) const{
+    if(points.size()!=4) throw runtime_error("Wrong Size in check_intersecting, Gale_diagram");
+    
+    vector<Point> positives;
+    vector<Point> negatives;
+    for(Point p : points){
+        if(p.get_sign()) positives.push_back(p);
+        else negatives.push_back(p);
+    }
+    int s = (int) positives.size();
```

Comment:
`size_t` is actually ok here, no need to do a hard cast like this.
Generally, avoid using hard casts, because they're nearly impossible to find via grep.
Use `static_cast` instead, as this can help you find problematic sections in your code later on